### ISSUE TITLE
VIP: optimizer notes

### ISSUE THREAD

**Issue by fubuloubu** [vyperlang/vyper#3337]
## Simple Summary
Add special comments to vyper code to inform the optimizer what additional optimizations it may consider "safe"

## Motivation
There are certain types of optimizations that one might make to Vyper code where understanding a global relationship would improve the ability of the optimizer to optimize code. For example, in an ERC20 implementation, typically there is a relationship that `totalSupply := sum(balanceOf)` for all possible addresses, but even this is a fairly complex relationship. We can actually reduce this to a much more actionable relationship for the purposes of optimization by saying that `totalSupply >= balanceOf`, which would tell the optimizer that any increments to both values by the same number should only throw once, so it's safe to perform one of them as unchecked arithmetic.

## Specification
```vyper
# @opt totalSupply >= balanceOf[*]
self.totalSupply += amount
self.balanceOf[receiver] += amount
```

## Backwards Compatibility
It's a comment, so no issues

## Dependencies
Probably requires that the optimizer allow extensible rules

## References
n/a

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/)

**Comment by fubuloubu** [comment]
Honestly, `totalSupply := sum(balanceOf)` is probably the more useful note to make, because it would satisfy for all of the following

```
# @opt totalSupply := sum(balanceOf)
# NOTE: Place at the top of the file
...

self.balanceOf[msg.sender] -= amount  # raises if overflow
self.balanceOf[receiver] += amount  # can be unchecked because `sum(balanceOf)` is unchanged
...

self.totalSupply += amount  # raises if overflow
self.balanceOf[receiver] += amount  # can be unchecked because `totalSupply == sum(balanceOf) <= amount`
...

self.balanceOf[msg.sender] -= amount  # raises if underflow
self.totalSupply -= amount  # can be unchecked because `sum(balanceOf) == totalSupply >= amount`
```

**Comment by fubuloubu** [comment]
I guess this gets into symbolic execution a little

**Comment by charles-cooper** [comment]
i think user changes to the optimizer can be more dangerous than inline assembly, since you don't really know how the compiler will interpret the rules

**Comment by fubuloubu** [comment]
Yes, this is definitely true. I think specifying invariants is a really nice capability to be able to specify it in natspec to pull it out would be great. above and beyond that, perhaps the optimizer is able to leverage invariants to craft better code but it does seem to increase the unpredictable behavior of the compiler
